File: TypeScript/examples.ts
function filter<T, S extends T>(
  array: T[],
  callbackfn: (value: T) => value is S
): S[] {
  const result: S[] = [];
  for (const value of array) {
    if (callbackfn(value)) {
      result.push(value);
    }
  }
  return result;
}

function flatten(l: any): string[] {
  if (Array.isArray(l)) {
    const t = flatten(l[0]).concat(flatten(l.slice(1)));
    return t;
  } else {
    if (typeof l === 'string') {
      return [l];
    } else {
      return [String(l)];
    }
  }
}

interface TreeNode {
    value: number;
    children?: TreeNode[];
    // Recursive reference to the same type
}

function isTreeNode(node: any): node is TreeNode {
  if (typeof node !== 'object' || node === null) {
    return false;
  }

  if (typeof node.value !== 'number') {
    return false;
  }

  if (node.children) {
    if (!Array.isArray(node.children)) {
      return false;
    }

    // Recursively check each child
    for (const child of node.children) {
      if (!isTreeNode(child)) {
        return false;
      }
    }
  }
  return true;
}

function rainfall(weather_reports : object[]): number {
  let total = 0, count = 0;
  for (let day of weather_reports) {
    if ("rainfall" in day) {
      const val = day["rainfall"]
      if (typeof val === "number" && 0 <= val && val <= 999) {
        total += val;
        count += 1;
      }
    }
  }
  if (count > 0) {
    return total / count;
  } else {
    return 0;
  }
}

File: Pyright/examples.py
from typing import TypeIs, Callable, TypeVar, Union

T = TypeVar("T")
S = TypeVar("S")
def filter(l: list[T], predicate: Callable[[T], TypeIs[S]]) -> list[S]:
    result = []
    for element in l:
        if predicate(element):
            result.append(element)
    return result

type MaybeNestedList[T] = list[MaybeNestedList[T]] | T
def flatten(l: MaybeNestedList[T]) -> list[T]:
    if isinstance(l, list):
        if l:
            return flatten(l[0]) + flatten(l[1:])
        else:
            return []
    else:
        return [l]

type TreeNode = tuple[int, list[TreeNode]]

def is_tree_node(node: object) -> TypeIs[TreeNode]:
    if not (isinstance(node, tuple) and len(node) == 2):
        return False
    else:
        if not isinstance(node[0], int):
            return False
        else:
            if not isinstance(node[1], list):
                return False
            else:
                for child in node[1]:
                    if not is_tree_node(child):
                        return False
                return True

type JSON = Union[str, float, bool, None, list[JSON], dict[str, JSON]]
def rainfall(weather_reports: list[JSON]) -> float:
    total = 0
    count = 0
    for day in weather_reports:
        if isinstance(day, dict) and "rainfall" in day:
            val = day["rainfall"]
            if isinstance(val, float) and 0 <= val <= 999:
                total += val
                count += 1
    return total / count if count > 0 else 0
